### PR TITLE
Environment variable for setting the max history for an environment

### LINKED ISSUE #8367: Environment variable for max release history
Environment variable for setting the max history for an environment. This would remove need to set it on each upgrade with the `--history-max` flag. Similar to `TILLER_HISTORY_MAX` variable in Helm 2.

Output of `helm version`:
Helm 3

### PR DESCRIPTION
**Pull request by wawa0210**
**What this PR does / why we need it**:
Environment variable for max release history

**Special notes for your reviewer**:

fix #8367

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by wawa0210** [comment]
@hickeyma @marckhouzam 

I'm a little confused, if other modules (such as rollback) want to use this method, `historyMaxFromEnv()`, where is the best place to put it. Do you have any better ideas

**Comment by bacongobbler** [comment]
`pkg/cli` is used to read and set other environment variables. That might be a better place to put this code logic.

https://github.com/helm/helm/blob/9b90e848ca659e147cae66dde9c16dcda578f113/pkg/cli/environment.go#L105

**Comment by wawa0210** [comment]
> `pkg/cli` is used to read and set other environment variables. That might be a better place to put this code logic.
> 
> https://github.com/helm/helm/blob/9b90e848ca659e147cae66dde9c16dcda578f113/pkg/cli/environment.go#L105

Thanks for reply, done.

File: pkg/cli/environment.go [hunk 1]
```diff
@@ -56,6 +59,8 @@ type EnvSettings struct {
 	RepositoryCache string
 	// PluginsDirectory is the path to the plugins directory.
 	PluginsDirectory string
+	// HistoryMax is the name of the env var for setting max history.
+	HistoryMax int
```
Comment (hickeyma) [inline pkg/cli/environment.go:63]: Observation: I think for consistency with the arg `--history-max`, `HistoryMax` should remain but really `MaxHistory` would read better.

File: pkg/cli/environment.go [hunk 2]
```diff
@@ -68,6 +73,7 @@ func New() *EnvSettings {
 		RegistryConfig:   envOr("HELM_REGISTRY_CONFIG", helmpath.ConfigPath("registry.json")),
 		RepositoryConfig: envOr("HELM_REPOSITORY_CONFIG", helmpath.ConfigPath("repositories.yaml")),
 		RepositoryCache:  envOr("HELM_REPOSITORY_CACHE", helmpath.CachePath("repository")),
+		HistoryMax:       envIntOr("HISTORY_MAX", defaultMaxHistory),
```
Comment (hickeyma) [inline pkg/cli/environment.go:76]: HISTORY_MAX --> HELM_HISTORY_MAX

File: pkg/cli/environment.go [hunk 3]
```diff
@@ -102,6 +108,18 @@ func envOr(name, def string) string {
 	return def
 }
 
+func envIntOr(name string, def int) int {
+	if name == "" {
+		return def
+	}
+	envVal := envOr(name, "")
```
Comment (hickeyma) [inline pkg/cli/environment.go:115]: Might be safer to also pass the default value instead of ""

**Review by hickeyma — CHANGES_REQUESTED** [review]
Thanks for pushing the PR @wawa0210. Some small things to address.

File: pkg/cli/environment.go (hunk 3 above)
Comment (wawa0210) [reply pkg/cli/environment.go]: done.

File: pkg/cli/environment.go (hunk 2 above)
Comment (wawa0210) [reply pkg/cli/environment.go]: done.

File: pkg/cli/environment.go (hunk 1 above)
Comment (wawa0210) [reply pkg/cli/environment.go:63]: I am also hesitant about this, if ok, I can update to "MaxHistory"

😄😄😄

Comment (liuming-dev) [reply pkg/cli/environment.go:63]: I agree with @hickeyma, `MaxHistory` reads better.

File: pkg/cli/environment.go [hunk 4]
```diff
@@ -68,6 +73,7 @@ func New() *EnvSettings {
 		RegistryConfig:   envOr("HELM_REGISTRY_CONFIG", helmpath.ConfigPath("registry.json")),
 		RepositoryConfig: envOr("HELM_REPOSITORY_CONFIG", helmpath.ConfigPath("repositories.yaml")),
 		RepositoryCache:  envOr("HELM_REPOSITORY_CACHE", helmpath.CachePath("repository")),
+		HistoryMax:       envIntOr("HELM_HISTORY_MAX", defaultMaxHistory),
```
Comment (liuming-dev) [inline pkg/cli/environment.go:76]: `HistoryMax` → `MaxHistory`

File: pkg/cli/environment.go [hunk 5]
```diff
@@ -119,6 +137,7 @@ func (s *EnvSettings) EnvVars() map[string]string {
 		"HELM_KUBECONTEXT":   s.KubeContext,
 		"HELM_KUBETOKEN":     s.KubeToken,
 		"HELM_KUBEAPISERVER": s.KubeAPIServer,
+		"HELM_HISTORYMAX":    strconv.Itoa(s.HistoryMax),
```
Comment (liuming-dev) [inline pkg/cli/environment.go:140]: `HELM_HISTORYMAX ` → `HELM_MAXHISTORY`

File: pkg/cli/environment.go (hunk 1 above)
Comment (wawa0210) [reply pkg/cli/environment.go]: Deal, modified.  😄

**Review by hickeyma — CHANGES_REQUESTED** [review]
Sorry for the confusion here. My comment in https://github.com/helm/helm/pull/8369#discussion_r447449985 is an observation. I would prefer to change it to `MaxHistory` but I think because of the flag which cannot be changed until Helm 4, we should follow the naming for consistency.

@wawa0210 Do you mind changing it back?

**Comment by wawa0210** [comment]
> I would prefer to change it to `MaxHistory` but I think because of the flag which cannot be changed until Helm 4, we should follow the naming for consistency.

What we change is only the name of the environment variable, not the name of the command line flag~ Will this affect? If there is a negative impact, I can try to roll back

**Review by hickeyma — CHANGES_REQUESTED** [review]
Looking good @wawa0210. Can you fix the CI build as style issue in `pkg/cli/environment.go`.

**Comment by hickeyma** [comment]
I am ok with the naming now.

**Comment by wawa0210** [comment]
> I am ok with the naming now.

It seems that all problems have been fixed, can you help approve?

**Comment by marckhouzam** [comment]
@hickeyma do you think the new environment variable, since it is user-facing, should use the same name as the flag?  So HELM_HISTORY_MAX? I'm not convinced it is essential but I wanted to ask.

**Comment by wawa0210** [comment]
@hickeyma  @marckhouzam @liuming-dev 

All questions have been answered, can you help approve?

File: pkg/cli/environment.go [hunk 6]
```diff
@@ -119,6 +137,7 @@ func (s *EnvSettings) EnvVars() map[string]string {
 		"HELM_KUBECONTEXT":   s.KubeContext,
 		"HELM_KUBETOKEN":     s.KubeToken,
 		"HELM_KUBEAPISERVER": s.KubeAPIServer,
+		"HELM_MAX_HISTORY":   strconv.Itoa(s.MaxHistory),
```
Comment (hickeyma) [inline pkg/cli/environment.go:140]: Move just after `HELM_NAMESPACE` because comment applies to the kube vars

File: pkg/cli/environment.go [hunk 7]
```diff
@@ -56,6 +59,8 @@ type EnvSettings struct {
 	RepositoryCache string
 	// PluginsDirectory is the path to the plugins directory.
 	PluginsDirectory string
+	// MaxHistory is the name of the env var for setting max history.
```
Comment (hickeyma) [inline pkg/cli/environment.go:62]: MaxHistory is the name of the env var for setting max history. --> MaxHistory is the max release history maintained.

**Review by hickeyma — CHANGES_REQUESTED** [review]
Nearly there @wawa0210, thanks. A few small nits and need to add the variable and description to https://github.com/helm/helm/blob/master/cmd/helm/root.go#L52

File: pkg/cli/environment.go (hunk 5 above)
Comment (hickeyma) [reply pkg/cli/environment.go]: no, `HELM_MAX_HISTORY` instead

**Comment by hickeyma** [comment]
> do you think the new environment variable, since it is user-facing, should use the same name as the flag? So HELM_HISTORY_MAX? I'm not convinced it is essential but I wanted to ask.

Thought about it alright @marckhouzam and its something I deliberated over. I think it can remain as its a new variable and I would hope the flag would be changed in Helm 4 to be more descriptive.

**Comment by hickeyma** [comment]
I am still a little unconvinced that the env variable needs to be global in the code (i.e. in `pkg/cli`) as I think it will probably only be used by `upgrade` as it makes the most sense. I am however, open to the feedback from @bacongobbler and @marckhouzam in https://github.com/helm/helm/pull/8369#issuecomment-651420879 and https://github.com/helm/helm/issues/8367#issuecomment-651041514.

**Comment by hickeyma** [comment]
@wawa0210 If you can updated this asap, we can try and get it into the 3.3.0 release. It is dues the RC today I think.
/cc @bacongobbler

**Comment by marckhouzam** [comment]
> I am still a little unconvinced that the env variable needs to be global in the code (i.e. in pkg/cli) as I think it will probably only be used by upgrade as it makes the most sense.

@hickeyma I guess it depends on if we want to apply the Max history logic to `helm rollback` too?  My impression is yes, but I can't evaluate if there are ramifications to doing that.

**Comment by bacongobbler** [comment]
we're cutting rc.1 later this morning, so we'll have to move this over to the 3.4.0 milestone.

**Comment by bacongobbler** [comment]
>  I guess it depends on if we want to apply the Max history logic to `helm rollback` too?

Yes, this should also be made available for `helm rollback` as it was for Helm 2.

File: pkg/cli/environment.go (hunk 6 above)
Comment (wawa0210) [reply pkg/cli/environment.go]: Feeling your feedback, done.

**Comment by wawa0210** [comment]
> @wawa0210 If you can updated this asap, we can try and get it into the 3.3.0 release. It is dues the RC today I think.

Sorry, I didn't have time to update yesterday, maybe the time is too late to 3.3.0 release

**Review by hickeyma — CHANGES_REQUESTED** [review]
It is nearly there @wawa0210. Do you mind adding a unit test in https://github.com/helm/helm/blob/master/pkg/cli/environment_test.go to test for the variable?

**Comment by hickeyma** [comment]
> Sorry, I didn't have time to update yesterday, maybe the time is too late to 3.3.0 release

That's ok. It can go into 3.4.

**Comment by wawa0210** [comment]
> It is nearly there @wawa0210. Do you mind adding a unit test in https://github.com/helm/helm/blob/master/pkg/cli/environment_test.go to test for the variable?

Added, troublesome review again

**Review by hickeyma — APPROVED** [review]
LGTM, thanks @wawa0210

**Comment by pchico83** [comment]
is this feature included in 3.3.0? Will be included on the next release?

**Comment by marckhouzam** [comment]
@pchico83 this feature is not in the 3.3 release but will be part of 3.4

**Pull request merged**